src/levanter/models/factorized_llama.py

```diff
@@ -0,0 +1,627 @@
+            hidden_dim=hf_config.hidden_size,
+            intermediate_dim=hf_config.intermediate_size,
+            num_layers=hf_config.num_hidden_layers,
+            num_heads=hf_config.num_attention_heads,
+            num_kv_heads=hf_config.num_key_value_heads,
+            activation_function=hf_config.hidden_act,
+            initializer_range=hf_config.initializer_range,
+            layer_norm_epsilon=hf_config.rms_norm_eps,
+            rope_scaling=hf_config.rope_scaling,
+            factor_dim=getattr(hf_config, "factor_dim", 512),
+        )
+
+    def to_hf_config(self, vocab_size: int, config_overrides: Optional[Dict] = None) -> HfLlamaConfig:
+        """Convert to HuggingFace's FactorizedLlamaConfig
+
+        Args:
+            vocab_size (int, optional): Vocabulary size of the tokenizer. Defaults to 32000.
+            config_overrides (dict, optional): Overrides for the config. Defaults to None.
+
+        Returns:
+            HfLlamaConfig: HuggingFace's FactorizedLlamaConfig
+        """
+        if config_overrides is None:
+            config_overrides = {}
+
+        return HfLlamaConfig(
+            max_position_embeddings=self.seq_len,
+            hidden_size=self.hidden_dim,
+            intermediate_size=self.intermediate_dim,
+            num_hidden_layers=self.num_layers,
+            num_attention_heads=self.num_heads,
+            num_key_value_heads=self.num_kv_heads,
+            hidden_act=self.activation_function,
+            initializer_range=self.initializer_range,
+            rms_norm_eps=self.layer_norm_epsilon,
+            rope_scaling=self.rope_scaling,
+            factor_dim=self.factor_dim,
+            vocab_size=vocab_size,
+            **config_overrides,
+        )
+
+    @property
+    def model_type(cls) -> Type["FactorizedLlamaLMHeadModel"]:
+        return FactorizedLlamaLMHeadModel
+
+
+def low_rank_approximation(matrix, rank):
+    """
+    Approximates the input matrix using Singular Value Decomposition (SVD) to a lower rank.
+
+    Args:
+        matrix (jax.numpy.ndarray): Input matrix of shape (N, M)
+        rank (int): Desired rank of the approximation (H)
+
+    Returns:
+        jax.numpy.ndarray, jax.numpy.ndarray: Two matrices of shape (N, H) and (H, M)
+    """
+    from jax.numpy.linalg import svd
+
+    # Perform SVD
+    U, S, Vh = svd(matrix, full_matrices=False)
+
+    S = S[..., :rank]
+    # Truncate U, S, and Vh to the desired rank
+    U_truncated = U[..., :rank]
+    if len(S.shape) == 1:
+        S_truncated = jnp.diag(S)
+    else:
+        S_truncated = jax.vmap(jnp.diag, 0)(S[..., :rank])
+
+    Vh_truncated = Vh[..., :rank, :]  # Note: Vh is already the conjugate transpose of V
+
+    # Reconstruct the low-rank approximation
+    down_proj = U_truncated @ S_truncated  # Shape (N, H)
+    up_proj = Vh_truncated  # Shape (H, M)
+
+    return down_proj, up_proj
+
+
+class FactorizedLinear(StateDictSerializationMixin, eqx.Module):
```

Comment:
so one thing you could do is do all this as a tree transformation/model surgery, similar to how we do it for Lora.

Basically you load the model as normal, then you do something like:


```python

def replace_linear(x):
    if not isinstance(x, Linear): return x
    
    up_proj, down_proj = low_rank_approximation(x)
    return LowRankLinear(up_proj, down_proj)

modified_model = jax.tree_util.tree_map(replace_linear, model, is_leaf=lambda x: isinstance(x, Linear))
```

I'm pretty sure you could delete most of this file if you did this.

This version only works if use_scan_layers is false. You have to do some fanciness (that we do in Lora) for scan layers to essentially vmap the replace_linear layer whenever you detect a stacked. (Maybe I should make a stacked_aware_tree_map or something)

Reply:
Oh that's a great idea. You'd lose a little flexibility in the size of your low-rank layers, but it would be nice to work across all models.

I'll try that out once I have something working end-to-end. (I don't mind starting with this and throwing it away, since it's a bit easier to debug this way than also debugging any wonky transform logic I try to write...